File: Reqnroll/Chapter16/16.1-ScenarioResourcePattern/Before/WIMP.Specs/StepDefinitions/PaymentsStepDefinitions.cs
using Reqnroll;

using WIMP.App.Models;
using WIMP.Specs.Drivers;
using WIMP.Specs.Support;

namespace WIMP.Specs.StepDefinitions;

[Binding]
public class PaymentsStepDefinitions(OrderingContext orderingContext, AppHostingContext appHostingContext, PaymentApiDriver paymentApiDriver)
{
    private Payment? payment;

    [When("their payment is authorised by the payment gateway")]
    public async Task WhenTheirPaymentIsAuthorisedByThePaymentGateway()
    {
        int orderNo = orderingContext.PlacedOrderNo ?? throw new InvalidOperationException("No order placed.");
        payment = await paymentApiDriver.AuthorizePaymentFor(orderNo).Execute();
    }

    [Then("they should be shown a success message with the payment reference")]
    public void ThenTheyShouldBeShownASuccessMessageWithThePaymentReference()
    {
        Assert.IsNotNull(payment, "No payment was authorized");
        Assert.IsTrue(payment.Success, "The payment should be successful.");
        Assert.IsNotEmpty(payment.PaymentReference, "The response should contain a payment reference");
        Assert.IsNotNull(appHostingContext.PaymentGateway?.GetPaymentByReference(payment.PaymentReference), "Payment reference is invalid");
    }
}

File: CucumberJVM/Chapter20/20.3-ParameterizedExecutionPattern/Before/src/test/java/com/wimp/app/specs/stepdefinitions/PaymentsStepDefinitions.java
package com.wimp.app.specs.stepdefinitions;

import com.wimp.app.models.Payment;
import com.wimp.app.specs.drivers.PaymentApiDriver;
import com.wimp.app.specs.drivers.PaymentGatewayScenarioContextProxy;
import com.wimp.app.specs.support.OrderingContext;
import io.cucumber.java.en.*;

import static org.junit.jupiter.api.Assertions.*;

public class PaymentsStepDefinitions {
    private final OrderingContext orderingContext;
    private final PaymentApiDriver paymentApiDriver;
    private final PaymentGatewayScenarioContextProxy paymentGatewayScenarioContextProxy;
    private Payment payment;

    public PaymentsStepDefinitions(OrderingContext orderingContext, PaymentApiDriver paymentApiDriver, PaymentGatewayScenarioContextProxy paymentGatewayScenarioContextProxy) {
        this.orderingContext = orderingContext;
        this.paymentApiDriver = paymentApiDriver;
        this.paymentGatewayScenarioContextProxy = paymentGatewayScenarioContextProxy;
    }

    @When("their payment is authorised by the payment gateway")
    public void theirPaymentIsAuthorisedByThePaymentGateway() throws Exception {
        int orderNo = orderingContext.getCurrentOrderNoVerified();
        payment = paymentApiDriver.authorizePaymentFor(orderNo).execute();
    }

    @Then("they should be shown a success message with the payment reference")
    public void theyShouldBeShownASuccessMessageWithThePaymentReference() {
        assertNotNull(payment, "No payment was authorized");
        assertTrue(payment.isSuccess(), "The payment should be successful.");
        assertFalse(payment.getPaymentReference().isEmpty(), "The response should contain a payment reference");
        assertNotNull(paymentGatewayScenarioContextProxy.getPaymentGatewayOverride(), "No simulated payment gateway available");
        assertNotNull(paymentGatewayScenarioContextProxy.getPaymentGatewayOverride().getPaymentByReference(payment.getPaymentReference()), "Payment reference is invalid");
    }
}
